test/grpc_test.rb

```diff
@@ -0,0 +1,215 @@
+    request = EchoMsg.new
+    open_circuit!(stub, :a_server_streaming_rpc, request)
+
+    assert_raises GRPC::CircuitOpenError do
+      stub.a_server_streaming_rpc(request)
+    end
+  end
+
+  def test_circuit_breaker_on_bidi_streamer
+    stub = build_insecure_stub(EchoStub, host: "0.0.0.1:0", opts: {semian_options: SEMIAN_OPTIONS})
+    requests = [EchoMsg.new, EchoMsg.new]
+    open_circuit!(stub, :a_bidi_rpc, requests)
+
+    assert_raises GRPC::CircuitOpenError do
+      stub.a_bidi_rpc(requests)
+    end
+  end
+
+  def test_bulkheads_tickets_are_working
+    run_services_on_server(@server, services: [EchoService]) do
+      stub1 = build_insecure_stub(EchoStub, opts: {semian_options: SEMIAN_OPTIONS})
+      stub1.semian_resource.acquire do
+        testing2_semian_options = {
+          name: :testing2,
+          tickets: 1,
+          timeout: 0,
+          error_threshold: ERROR_THRESHOLD,
+          success_threshold: 1,
+          error_timeout: 10,
+        }
+        stub2 = build_insecure_stub(EchoStub, opts: {semian_options: testing2_semian_options})
+        stub2.semian_resource.acquire do
+          assert_raises GRPC::ResourceBusyError do
+            stub2.an_rpc(EchoMsg.new)
+          end
+        end
+      end
+    end
+  end
+
+  private
+
+  def open_circuit!(stub, method, args)
+    ERROR_THRESHOLD.times do
+      assert_raises GRPC::Unavailable do
+        stub.send(method, args)
+      end
+    end
+  end
+
+  def build_insecure_stub(klass, host: nil, opts: nil)
+    host ||= @host
+    opts ||= @client_opts
+    klass.new(host, :this_channel_is_insecure, **opts)
+  end
+
+  def build_rpc_server(server_opts: {}, client_opts: {})
+    @hostname = '0.0.0.0'
+    @server = new_rpc_server_for_testing({poll_period: 1}.merge(server_opts))
+    @port = @server.add_http2_port("#{@hostname}:0", :this_port_is_insecure)
+    @host = "#{@hostname}:#{@port}"
+    @client_opts = client_opts
+    populate_toxiproxy
+    @server
+  end
+
+  def populate_toxiproxy
+    Toxiproxy.populate([
+      {
+        name: 'semian_test_grpc',
+        upstream: "#{@hostname}:#{@port}",
+        listen: "#{@hostname}:#{@port + 1}",
+      },
+    ])
+  end
+
+  def new_rpc_server_for_testing(server_opts = {})
+    server_opts[:server_args] ||= {}
+    update_server_args_hash(server_opts[:server_args])
+    GRPC::RpcServer.new(**server_opts)
+  end
+
+  def update_server_args_hash(server_args)
+    so_reuseport_arg = 'grpc.so_reuseport'
+    unless server_args[so_reuseport_arg].nil?
+      fail 'Unexpected. grpc.so_reuseport already set.'
+    end
+    # Run tests without so_reuseport to eliminate the chance of
+    # cross-talk.
+    server_args[so_reuseport_arg] = 0
+  end
+
+  def run_services_on_server(server, services: [])
+    services.each do |s|
+      server.handle(s)
+    end
+    t = Thread.new { server.run }
+    server.wait_till_running
+
+    yield
+
+    server.stop
+    t.join
```

Comment:
What happens when `yield` throws an exception? Do we leak server threads?

Reply:
I took that part from the GRPC test helper: https://github.com/grpc/grpc/blob/master/src/ruby/spec/support/helpers.rb

What do you think I should do to make sure we won't leak server threads?

Reply:
Stick an ensure after the yield. IMO. @pushrax agree?

Reply:
Plus a `begin` before the thread allocation 👍